Updating the "How to Seednode" docs (#148)
bisq-network/bisq-docs: exchange/howto/run-seednode.adoc

== What is a Bisq seed node

Bisq uses a highly anonymous and dynamic P2P network for communications between Bisq clients. In order to get access to the network, the Bisq network features always-up Bisq clients with a static addresses: seed nodes. They have 2 main tasks:

. Deliver the peer addresses of other dynamic network peers

. Deliver the initial P2P network data like open offers, mailbox messages and arbitrators

Being the gateways to the Bisq P2P network, seed nodes are among the first nodes a Bisq client connects to on each startup. Therefore, seed nodes have distinct properties:

- Seed nodes face a heavy load and are therefore operated on servers offering a high level of reliability and better connectivity than a normal Bisq client.
- A bad seed node can have a substantial impact on the Bisq network. A Bisq user may encounter anything from a longer startup time of her Bisq client (which is unconvenient but not critical) to missing messages sent to her by a trading partner or arbitrator while her Bisq client has been offline (which can lead to disputes, failed trades, blocking of the Bisq client and bad reputation of the Bisq client and therefore the user). 

> All in all, seed nodes are important for the Bisq network to function in a user-friendly way. Without the seed nodes, a fresh Bisq client would not find its way into the network, at least not without manual intervention. Bad seed nodes can make it harder for a fresh client to join the network. In the end, seed nodes allow new users easy access to the services of Bisq and should therefore be kept up and running.

Reviewer: ```suggestion
All in all, seed nodes play a vital role in the Bisq network. They allow users easy access to the Bisq network and therefore must be kept online and available at all times. Without the seed nodes, a new Bisq client would not find its way into the network, at least not without manual intervention. Bad seed nodes can make it harder for a new client to join the network.
```

== Seed node operators

Being a seed node operator brings along duties and benefits making it worthwhile.

=== Duties of the seed node operator

The operator of a seed node needs to:

> - find yourself a deputy in case you are on not available (travel, holiday, others, ...). Usually, on of the developers is there to help. Just inform them on your setup and all is good.

Reviewer: ```suggestion
- find yourself a deputy in case you are not available (travel, holiday, others, ...). Usually, one of the developers is there to help. Just inform them on your setup and all is good.
```

- keep the server up and running, including patching and hardening the operating system
- keep the seed node software up to date

> - keep an eye on the Bisq's communication channels for updates, issues, countermeasures,... (see <<Getting in Touch>>)

Reviewer: ```suggestion
- keep an eye on Bisq's communication channels for updates, issues, countermeasures,... (see <<Getting in Touch>>)
```

- keep an eye on the logs of the seed node and the operating system, and act if something shows up
- if necessary, report and escalate (see <<Getting in Touch>>)

> - set up a bond of 2000 BSQ to get the privilege to run a seed node. In case of severe failure of service, the bond gets confiscated (burned).

Author: you sure? I took the 2000 from the previous version (see line 105 of the original howto). If 20000 is the correct number, do you have any reference to back it up?

> - set up a bond of 2000 BSQ to get the privilege to run a seed node. In case of severe failure of service, the bond gets confiscated (burned).

Reviewer: The original document seems to be outdated. Currently the BONDED_ROLE_FACTOR is 1000 and the SEED_NODE_OPERATOR has a multiplier of 20, giving a bonded amount of 20000.
Reference:
- https://github.com/bisq-network/bisq/blob/master/core/src/main/java/bisq/core/dao/governance/param/Param.java#L125
- https://github.com/bisq-network/bisq/blob/master/core/src/main/java/bisq/core/dao/state/model/governance/BondedRoleType.java#L57

[NOTE]
====
With the advent of the DAO and the subsequent need of another 400GB of ROM, the numbers here are currently under discussion. Until new figures are decided on, state your expenses for running the server and add the 50 BSQ on top to get your sum for compensation.
====

== Getting in Touch

Bisq uses Slack to communicate via chat. As a seed node operator, you are required to

- subscribe to the Bisq Slack channel `bisq-seednode` and `bisq-monitor`
- the `bisq-seednode` channel is the place where updates, issues, countermeasures, heads-ups, ... are discussed. If you encounter a problem with your seed node and cannot solve it by yourself, this is the place to report to (with a specific question, logs, ...). A developer will get back to you.
- the `bisq-monitor` channel is the place where issues with seed nodes are reported, either manually or by our monitoring service. If your seed node is mentioned for having an issue, you are required to react.
- please be responsive when addressed

== System requirements for hosting machine

- Min. 4 GB of RAM
- 400 GB of disk space
- 2 TB network traffic
- Uptime of > 99.9%
- > 10000 file handles available (check `ulimit -Sn`)

> And please try to stay away from cheap hosting providers! They tend to not care about QOS, the hardware is slow, the occasionally even turned a service off without warning - ask us how we know...

Reviewer: ```suggestion
And please try to stay away from cheap hosting providers! They tend to not care about QOS, the hardware is slow, they occasionally even turned a service off without warning - ask us how we know...
```

## Get a seed node up and running

> You can run your seed node in many ways. The guide you are currently reading applies to a (debian-based) unix system with systemd available. Furthermore, we assume that only one seed node is operated per host. Please be aware that the guide is only a guide and not a step by step copy and paste how to, as we ignore for example the need for the occational root access and file permissions, swap, general opsec, among other things. In case your host differs from our reference system, we are sure you can handle yourself.

Reviewer: ```suggestion
You can run your seed node in many ways. The guide you are currently reading applies to a (debian-based) unix system with systemd available. Furthermore, we assume that only one seed node is operated per host. Please be aware that the guide is only a guide and not a step by step copy and paste how to, as we ignore for example the need for the occasional root access and file permissions, swap, general opsec, among other things. In case your host differs from our reference system, we are sure you can handle yourself.
```

Author: What exactly is your suggestion here?

Reviewer: For some reason it is not picking up my diff. I was suggesting to change "occational" -> "occasional".

Although every port that is to be used only by localhost clients should be hardened by configuration and source code already, we recommend whitelisting stuff that is actually needed. Here is an example on how to configure iptables accordingly (the config drops any incoming connections except connection from localhost, SSH and responses.):

	iptables -P INPUT DROP && iptables -A INPUT -i lo -j ACCEPT && iptables -A INPUT -p tcp --dport 22 -j ACCEPT && iptables -A INPUT -m state --state ESTABLISHED,RELATED -j ACCEPT

[NOTE]
====
Make sure you execute *all `iptables` commands at once*! Executing only the first one is going to kick you out of your ssh session and does not let you back in.
====

=== Bitcoin Node

Start by setting up a user for the bitcoin core stuff.

	useradd -r -m bitcoind
	cd /home/bitcoind

Download the Bitcoin Core binaries https://bitcoin.org/en/download[here] to this directory, unpack it and, for updating convenience, create a symlink to the necessary binaries (so you only have to change the symlink on update):

	tar xzf bitcoin-0.18.0-x86_64-linux-gnu.tar.gz
	ln -s bitcoin-0.18.0/bin/bitcoind bitcoind
	ln -s bitcoin-0.18.0/bin/bitcoin-cli bitcoin-cli

> Create a systemd service file `bitcoind.service` in the systemd service path or your operating system (something like `/usr/lib/systemd/system/`) and adapt it to your needs. We recommend to create a user `bitcoind` for service hardening reasons (). In the end, it should look like

Reviewer: ```suggestion
Create a systemd service file `bitcoind.service` in the systemd service path for your operating system (something like `/usr/lib/systemd/system/`) and adapt it to your needs. We recommend to create a user `bitcoind` for service hardening reasons. In the end, it should look like
```

----
[Unit]
Description=Bitcoind
After=network.target

[Service]
ExecStart=bitcoind -daemon \

>                    -printtoconsole \

Reviewer: Is this desired?

Author: what exactly do you mean?

Reviewer: I am just not sure if it is necessary to printtoconsole for a daemon process.

Author: ah, well. The printtoconsole option sends outputs to stderr. On a system running systemd stderr ends up in the systems journal accessible via journalctl.

                   -nodebuglogfile \
                   -pid=/home/bitcoind/.bitcoin/bitcoind.pid
ExecStop=/home/bitcoind/bitcoin-cli stop

Type=forking
PIDFile=/home/bitcoind/.bitcoin/bitcoind.pid
Restart=on-failure

User=bitcoind
Group=bitcoind

PrivateTmp=true
ProtectSystem=full
NoNewPrivileges=true
PrivateDevices=true
MemoryDenyWriteExecute=true

[Install]
WantedBy=multi-user.target
----

Create a file `/home/bitcoind/.bitcoin/bitcoin.conf` that contains

----
server=1
txindex=1

> dbcache=1000

Reviewer: I am not familiar nor sure if this is necessary as none of my nodes specify this.

Author: I took these as input from the discussion on slack (bisq-seednode around 2019-07-21 noon UTC). May we have your options @sqrrm @mrosseel?

Author: I aligned the config to my working development config and removed the performance tweaks as well. I am leaving the `txindex=1` as it has been reported as needed by @sqrrm via slack.

maxconnections=800
timeout=30000
listen=0
rpcallowip=127.0.0.1
rpcuser=YOUR_USER_NAME
rpcpassword=YOUR_PW
blocknotify=sh ~/.bitcoin/blocknotify.sh %s
----

and another file `/home/bitcoind/.bitcoin/blocknotify.sh` that contains

  #!/bin/sh
  echo $1 | nc -w 1 127.0.0.1 5120

and make it executable (`chmod +x /home/bitcoind/.bitcoin/blocknotify.sh`).

Finally, enable and start the service

  systemctl enable bitcoind.service
  systemctl start bitcoind.service

and observe the logs

  journalctl --unit bitcoind --follow

> an check if anything works as expected.

Reviewer: ```suggestion
and check if everything works as expected.
```

=== Bisq Seed Node

Start by getting OpenJDK 10 up and running.

For getting the Bisq binaries, we recommend cloning the Bisq Git repository and compiling the code on your server. This way, you have precise control over what version you want to deploy. Furthermore, updating is very simple, just pull the changes, recompile and restart your service.

Furthermore, we recommend creating a user `bisq` in group `bisq` for service hardening reasons and using the `bisq`-users home directory to:

  useradd -r -m bisq
  cd /home/bisq
  git clone git@github.com:bisq-network/bisq.git
  cd bisq
  ./gradlew build -x test

> Create a systemd service file `bisq-seednode.service` (or copy the one shipped with bisq `$bisqdir/seednode/bisq-seednode.service`) in the systemd service path or your operating system (something like `/usr/lib/systemd/system/`) and adapt it to your needs.

Reviewer: ```suggestion
Create a systemd service file `bisq-seednode.service` (or copy the one shipped with bisq `$bisqdir/seednode/bisq-seednode.service`) in the systemd service path for your operating system (something like `/usr/lib/systemd/system/`) and adapt it to your needs.
```

In the end, your file should look something like

----
[Unit]
Description=Bisq Seed Node
After=network.target

[Service]
Environment="JAVA_OPTS=-Xms512M -Xmx2000M -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=6969 -Dcom.sun.management.jmxremote.rmi.port=6969 -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.authenticate=false"

> ExecStart=/home/bisq/bisq/bisq-seednode --appName=bisq-seednode --nodePort=8000 --userDataDir=/home/bisq/ --maxConnections=50 --fullDaoNode=true --rpcUser=YOUR_USER_NAME --rpcPassword=YOUR_PW --rpcPort=8332 --rpcBlockNotificationPort=5120

Reviewer: Not sure if maxConnections 50 is too high. Typically we use 30.

Author: Back when I ran 2 seed nodes, I had it set to 50 with only 800MB of Xmx and it worked without running out of memory. How about changing it to 30 and then adjust once we have a continuous heap space reading of the seed nodes?

Reviewer: Yes I think that makes sense. I am willing to be a guinea pig and optimize these settings on one seed once I add monitoring to it.

Restart=on-failure

User=bisq
Group=bisq

PrivateTmp=true
ProtectSystem=full
NoNewPrivileges=true
PrivateDevices=true
MemoryDenyWriteExecute=true

[Install]
WantedBy=multi-user.target
----

Note that the jmxremote JVM arguments are later used for monitoring the service, the rpc arguments are there to get the seed node hooked to the bitcoin service. Make sure that the YOUR_USER_NAME and YOUR_PW placeholders match the configuration of <<Bitcoin Node>>.

Enable and start the seed node by

`systemctl daemon-reload` +
`systemctl enable bisq-seednode.service` +
`systemctl start bisq-seednode.service`

> Keep an eye on the logs and see if anything works as expected:

Reviewer: ```suggestion
Keep an eye on the logs and see if everything works as expected:
```

=== System health reports

Since seed nodes are such a crucial part of the Bisq network, we require periodic health reports to our https://monitor.bisq.network[monitor]. Since the monitor only accepts plain TCP connections for incoming data, we have to accept a bit of overhead to keep the monitor from being flooded with unauthorized input.

In order to successfully report to the monitor, we need to create a TCP reverse proxy local to your host that can authenticate to the monitor. For this guide, we go with nginx, if you prefer another reverse proxy, we are sure you can handle yourself.

First of all, if you have not already, install nginx on your system.

Then proceed to creating the SSL certificate that is later used to authenticate against the monitor:

  cd /etc/nginx
  openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /etc/nginx/cert.key -out /etc/nginx/cert.crt

Use `ON = bisq.network`, `OU = seednodes` and `CN = <your seed nodes onion address here (without the ".onion" part)>` for certificate creation. The onion address can be found in the `hostname`-file mentioned before.

> Configure the reverse proxy with clientssl enabled. You can simply append the snipped below to your `/etc/nginx/nginx.conf` file:

Reviewer: ```suggestion
Configure the reverse proxy with clientssl enabled. You can simply append the snippet below to your `/etc/nginx/nginx.conf` file:
```

----
stream {
	log_format basic '$remote_addr [$time_local] '
	                 '$protocol Status $status Sent $bytes_sent Received $bytes_received '
	                 'Time $session_time';

	error_log syslog:server=unix:/dev/log;
	access_log syslog:server=unix:/dev/log basic;

	server {
		listen 2003;
		allow 127.0.0.1;
		deny all;
		proxy_pass monitor.bisq.network:2003;
		proxy_ssl on;

		proxy_ssl_certificate /etc/nginx/cert.crt;
		proxy_ssl_certificate_key /etc/nginx/cert.key;

		proxy_ssl_session_reuse on;
	}
}
----

> Start your nginx and observe the logs to see if anything works as expected:

Reviewer: ```suggestion
Start your nginx and observe the logs to see if everything works as expected:
```

> == Introduction

Reviewer 2: AsciiDoc automatically makes the first line into a subtitle of sorts, so I would suggest replacing lines 5-13 with the following. It includes minor edits in spacing, commas, etc (no material changes).
```
Bisq seed nodes serve as static gateways for Bisq clients to get connected to Bisq's dynamic P2P network.

In this doc, you will find information about:

- what a seed node is and how it works
- the duties of a seed node operator
- the benefits of being a seed node operator
- how to get a seed node up and running
- where to get in touch
```

Author: aggreed and done

All in all, seed nodes play a vital role in the Bisq network. They allow users easy access to the Bisq network and therefore must be kept online and available at all times. Without the seed nodes, a new Bisq client would not find its way into the network, at least not without manual intervention. Bad seed nodes can make it harder for a new client to join the network.

== Seed node operators

Being a seed node operator brings along duties and benefits making it worthwhile.

=== Duties of the seed node operator

The operator of a seed node needs to:

- find yourself a deputy in case you are not available (travel, holiday, others, ...). Usually, one of the developers is there to help. Just inform them on your setup and all is good.
- keep the server up and running, including patching and hardening the operating system
- keep the seed node software up to date
- keep an eye on Bisq's communication channels for updates, issues, countermeasures,... (see <<Getting in Touch>>)
- keep an eye on the logs of the seed node and the operating system, and act if something shows up
- if necessary, report and escalate (see <<Getting in Touch>>)

> - set up a bond of 20000 BSQ to get the privilege to run a seed node. In case of severe failure of service, the bond gets confiscated (burned).

Author: I disagree as different locales handle these separators differently. In large portions of Europe, for example, the default number format is 20.000,00 for "twenty-thousand point zero". Yes, the text is English but imho this is more confusing than not adding any aid. I add the number spelled in words to clarify.

- create a monthly report https://github.com/bisq-network/roles/issues/15[here]

> === Payment

Reviewer 2: ```suggestion
=== Compensation
```
The term 'compensation' clearer and more consistent with other docs, I think.

A seed node operator has the right to:

- file a compensation request over 200 BSQ for setting up a seed node
- file a monthly compensation request. We define 50 BSQ per month as an appropriate compensation per seed node per month.

[NOTE]
====
With the advent of the DAO and the subsequent need of an additional 300GB of disk space for the Bitcoin full node, the numbers here are currently under discussion. Until new figures are decided on, state your expenses for running the server and add the 50 BSQ on top to get your sum for compensation.
====

== Getting in Touch

Bisq uses Slack to communicate via chat. As a seed node operator, you are required to

- subscribe to the Bisq Slack channel `bisq-seednode` and `bisq-monitor`

> - the `bisq-seednode` channel is the place where updates, issues, countermeasures, heads-ups, ... are discussed. If you encounter a problem with your seed node and cannot solve it by yourself, this is the place to report to (with a specific question, logs, ...). A developer will get back to you.

Reviewer 2: ```suggestion
- the `bisq-seednode` channel is the place where updates, issues, countermeasures, heads-ups, etc are discussed. If you encounter a problem with your seed node and cannot solve it by yourself, this is the place to report to (with a specific question, logs, etc). A developer will get back to you.
```

- the `bisq-monitor` channel is the place where issues with seed nodes are reported, either manually or by our monitoring service. If your seed node is mentioned for having an issue, you are required to react.
- please be responsive when addressed

== System requirements for hosting machine

- Min. 4 GB of RAM
- 400 GB of disk space
- 2 TB network traffic
- Uptime of > 99.9%
- > 10000 file handles available (check `ulimit -Sn`)

And please try to stay away from cheap hosting providers! They tend to not care about QOS, the hardware is slow, they occasionally even turned a service off without warning - ask us how we know...

## Get a seed node up and running

> You can run your seed node in many ways. The guide you are currently reading applies to a (debian-based) unix system with systemd available. Furthermore, we assume that only one seed node is operated per host. Please be aware that the guide is only a guide and not a step by step copy and paste how to, as we ignore for example the need for the occasional root access and file permissions, swap, general opsec, among other things. In case your host differs from our reference system, we are sure you can handle yourself.

Reviewer 2: ```suggestion
You can run your seed node in many ways. The guide you are currently reading applies to a (debian-based) unix system with systemd available. Furthermore, we assume that only one seed node is operated per host. Please be aware that the guide is only a guide and not a step-by-step copy-and-paste how-to, as we ignore for example the need for the occasional root access and file permissions, swap, general opsec, among other things. In case your host differs from our reference system, we are sure you can handle yourself.
```

Group=bitcoind

PrivateTmp=true
ProtectSystem=full
NoNewPrivileges=true
PrivateDevices=true
MemoryDenyWriteExecute=true

[Install]
WantedBy=multi-user.target
----

Create a file `/home/bitcoind/.bitcoin/bitcoin.conf` that contains

----
server=1
txindex=1
dbcache=1000
maxconnections=800
timeout=30000
listen=0
rpcallowip=127.0.0.1
rpcuser=YOUR_USER_NAME
rpcpassword=YOUR_PW
blocknotify=sh ~/.bitcoin/blocknotify.sh %s
----

and another file `/home/bitcoind/.bitcoin/blocknotify.sh` that contains

  #!/bin/sh
  echo $1 | nc -w 1 127.0.0.1 5120

and make it executable (`chmod +x /home/bitcoind/.bitcoin/blocknotify.sh`).

Finally, enable and start the service

  systemctl enable bitcoind.service
  systemctl start bitcoind.service

and observe the logs

  journalctl --unit bitcoind --follow

and check if everything works as expected.

=== Bisq Seed Node

Start by getting OpenJDK 10 up and running.

For getting the Bisq binaries, we recommend cloning the Bisq Git repository and compiling the code on your server. This way, you have precise control over what version you want to deploy. Furthermore, updating is very simple, just pull the changes, recompile and restart your service.

> Furthermore, we recommend creating a user `bisq` in group `bisq` for service hardening reasons and using the `bisq`-users home directory to:

Reviewer 2: ```suggestion
Furthermore, we recommend creating a user `bisq` in group `bisq` for service hardening reasons and using the `bisq` user's home directory to:
```

(assuming this is the intended meaning)

  hostname
  private_key

to a secure location. In case your server loses the original files during a crash, you can recover easily by following the steps for taking over a seed node. All other data like the `db` or the `keys` directory are not relevant for the seed node.

=== System health reports

Since seed nodes are such a crucial part of the Bisq network, we require periodic health reports to our https://monitor.bisq.network[monitor]. Since the monitor only accepts plain TCP connections for incoming data, we have to accept a bit of overhead to keep the monitor from being flooded with unauthorized input.

In order to successfully report to the monitor, we need to create a TCP reverse proxy local to your host that can authenticate to the monitor. For this guide, we go with nginx, if you prefer another reverse proxy, we are sure you can handle yourself.

First of all, if you have not already, install nginx on your system.

Then proceed to creating the SSL certificate that is later used to authenticate against the monitor:

  cd /etc/nginx
  openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /etc/nginx/cert.key -out /etc/nginx/cert.crt

> Use `ON = bisq.network`, `OU = seednodes` and `CN = <your seed nodes onion address here (without the ".onion" part)>` for certificate creation. The onion address can be found in the `hostname`-file mentioned before.

Reviewer 2: ```suggestion
Use `ON = bisq.network`, `OU = seednodes` and `CN = <your seed nodes onion address here (without the ".onion" part)>` for certificate creation. The onion address can be found in the `hostname` file mentioned before.
```
